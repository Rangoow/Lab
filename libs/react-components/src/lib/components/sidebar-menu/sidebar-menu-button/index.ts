export * from './sidebar-menu-button.component';
