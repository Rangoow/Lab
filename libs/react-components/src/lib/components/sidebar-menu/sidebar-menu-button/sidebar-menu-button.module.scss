.button--collapsed {
  border-radius: 0;
}

:global(.vtmn-btn).button--uncollapsed {
    border-radius: 0;
    justify-content: start;
    padding-left: var(--vtmn-spacing_4);
}
