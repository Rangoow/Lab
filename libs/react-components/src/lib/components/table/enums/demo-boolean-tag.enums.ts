export enum EBooleanTags {
  TRUE = 'TRUE',
  FALSE = 'FALSE',
}
