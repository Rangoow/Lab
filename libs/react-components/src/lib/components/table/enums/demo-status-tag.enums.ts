export enum EDemoStatusTag {
  STATUS1 = 'STATUS 1',
  STATUS2 = 'STATUS 2',
  STATUS3 = 'STATUS 3',
}
