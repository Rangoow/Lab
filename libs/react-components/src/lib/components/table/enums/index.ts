export * from './demo-boolean-tag.enums';
export * from './demo-status-tag.enums';
