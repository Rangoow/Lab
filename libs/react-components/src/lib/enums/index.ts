export * from './react-components-data-test-id.enum';
export * from './react-components-illustrations-data-test-id.enum';
