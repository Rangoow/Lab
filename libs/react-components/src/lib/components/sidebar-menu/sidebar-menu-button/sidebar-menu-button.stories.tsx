import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';
import { SidebarMenuButton } from './sidebar-menu-button.component';

const meta: Meta<typeof SidebarMenuButton> = {
  title: 'Structure/Sidebar Menu Button',
  component: SidebarMenuButton,
  decorators: [
    // eslint-disable-next-line @typescript-eslint/naming-convention
    Story => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
};

type Story = StoryObj<typeof SidebarMenuButton>;

export default meta;

export const Collapsed: Story = {
  parameters: {
    backgrounds: {
      default: 'primary',
    },
  },
  args: {
    isCollapsed: true,
    icon: 'earth-line',
    text: 'Dashboard',
    title: 'Navigate to requests-dashboard',
    redirectPath: '/toto',
  },
};

export const UnCollapsed: Story = {
  parameters: {
    backgrounds: {
      default: 'primary',
    },
  },
  args: {
    isCollapsed: false,
    icon: 'earth-line',
    text: 'Dashboard',
    title: 'Navigate to requests-dashboard',
    redirectPath: '/toto',
  },
};
