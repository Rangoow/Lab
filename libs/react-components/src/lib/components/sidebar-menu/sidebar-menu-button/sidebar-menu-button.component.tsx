import { VitamixId } from '@vtmn/icons/dist/vitamix/font/vitamix';
import { VtmnButton, VtmnIcon } from '@vtmn/react';
import classNames from 'classnames/bind';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router-dom';
import { EReactComponentsDataTestId } from '../../../enums';
import styles from './sidebar-menu-button.module.scss';

interface IProps {
  readonly icon: VitamixId;
  readonly text: string;
  readonly title: string;
  readonly isCollapsed?: boolean;
  readonly redirectPath: string;
}

export function SidebarMenuButton({ icon, text, isCollapsed = false, redirectPath, title }: IProps) {
  const navigate = useNavigate();
  const { t } = useTranslation();
  const cx = classNames.bind(styles);

  return (
    <VtmnButton
      className={cx({ 'button--collapsed': isCollapsed }, { 'button--uncollapsed': !isCollapsed })}
      title={t(title)}
      variant="ghost"
      size="stretched"
      onClick={() => navigate(redirectPath)}
    >
      <VtmnIcon data-testid={EReactComponentsDataTestId.SIDEBAR_MENU_BUTTON_ICON} value={icon} />
      {isCollapsed ? null : <span className="typo_text-3 ml-4">{t(text)}</span>}
    </VtmnButton>
  );
}
