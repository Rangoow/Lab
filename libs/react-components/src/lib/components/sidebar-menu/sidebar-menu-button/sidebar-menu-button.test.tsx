import { RenderResult, screen } from '@testing-library/react';
import { UserEvent, userEvent } from '@testing-library/user-event';
import { VitamixId } from '@vtmn/icons/dist/vitamix/font/vitamix';
import * as router from 'react-router';
import { MockInstance } from 'vitest';
import { EReactComponentsDataTestId } from '../../../enums';
import { customRender } from '../../../tests/custom-render';
import { SidebarMenuButton } from './sidebar-menu-button.component';

const user: UserEvent = userEvent.setup();
const mockedUseNavigate = vi.fn();

describe(`uiSidebarMenuButton`, (): void => {
  let icon: VitamixId;
  let buttonText: string;
  let buttonTitle: string;
  let isButtonCollapsed: boolean;
  let buttonRedirectPath: string;
  let spyOnNavigate: MockInstance;

  describe('snapshot testing', (): void => {
    beforeEach((): void => {
      icon = `user-line`;
      buttonText = `user button`;
      buttonTitle = `navigate to user section`;
      buttonRedirectPath = `/test`;
    });

    describe(`when the menu is collapsed`, (): void => {
      beforeEach((): void => {
        isButtonCollapsed = true;
      });

      it(`should match the collapsed button snapshot`, async (): Promise<void> => {
        expect.assertions(1);

        const result: RenderResult = await customRender(
          <SidebarMenuButton
            icon={icon}
            text={buttonText}
            title={buttonTitle}
            isCollapsed={isButtonCollapsed}
            redirectPath={buttonRedirectPath}
          />,
        );

        expect(result).toMatchSnapshot();
      });
    });

    describe(`when the menu is not collapsed`, (): void => {
      beforeEach((): void => {
        isButtonCollapsed = false;
      });

      it(`should match the not collapsed button snapshot`, async (): Promise<void> => {
        expect.assertions(1);

        const result: RenderResult = await customRender(
          <SidebarMenuButton
            icon={icon}
            text={buttonText}
            title={buttonTitle}
            isCollapsed={isButtonCollapsed}
            redirectPath={buttonRedirectPath}
          />,
        );

        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('component testing', (): void => {
    beforeEach((): void => {
      icon = `user-line`;
      buttonText = `user button`;
      buttonTitle = `navigate to user section`;
      buttonRedirectPath = `/test`;
      spyOnNavigate = vi.spyOn(router, `useNavigate`).mockImplementation(() => mockedUseNavigate);
    });

    describe(`when the menu is collapsed`, (): void => {
      beforeEach((): void => {
        isButtonCollapsed = true;
      });

      it(`should only display the button with an icon and the title`, async (): Promise<void> => {
        expect.assertions(3);

        await customRender(
          <SidebarMenuButton
            icon={icon}
            text={buttonText}
            title={buttonTitle}
            isCollapsed={isButtonCollapsed}
            redirectPath={buttonRedirectPath}
          />,
        );
        const button: HTMLElement = screen.getByRole(`button`, { name: /navigate to user section/i });
        const buttonIcon: HTMLElement = screen.getByTestId(EReactComponentsDataTestId.SIDEBAR_MENU_BUTTON_ICON);

        expect(button).toBeVisible();
        expect(button).toHaveTextContent(``);
        expect(buttonIcon).toHaveClass(`vtmx-user-line`);
      });
    });

    describe(`when the menu is not collapsed`, (): void => {
      beforeEach((): void => {
        isButtonCollapsed = false;
      });

      it(`should display the button with an icon a text and the title`, async (): Promise<void> => {
        expect.assertions(3);

        await customRender(
          <SidebarMenuButton
            icon={icon}
            text={buttonText}
            title={buttonTitle}
            isCollapsed={isButtonCollapsed}
            redirectPath={buttonRedirectPath}
          />,
        );
        const button: HTMLElement = screen.getByRole(`button`, { name: /user button/i });
        const buttonIcon: HTMLElement = screen.getByTestId(EReactComponentsDataTestId.SIDEBAR_MENU_BUTTON_ICON);

        expect(button).toBeVisible();
        expect(button).toHaveAttribute(`title`, `navigate to user section`);
        expect(buttonIcon).toHaveClass(`vtmx-user-line`);
      });
    });

    describe(`when the user click on the button`, (): void => {
      it(`should called the button action function`, async (): Promise<void> => {
        expect.assertions(1);

        await customRender(
          <SidebarMenuButton
            icon={icon}
            text={buttonText}
            title={buttonTitle}
            isCollapsed={isButtonCollapsed}
            redirectPath={buttonRedirectPath}
          />,
        );

        const button: HTMLElement = screen.getByRole(`button`);

        await user.click(button);

        expect(spyOnNavigate).toHaveBeenCalledOnce();
      });
    });
  });
});
