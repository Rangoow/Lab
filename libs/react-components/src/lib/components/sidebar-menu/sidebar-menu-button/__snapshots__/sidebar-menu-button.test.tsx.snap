// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uiSidebarMenuButton > snapshot testing > when the menu is collapsed > should match the collapsed button snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--collapsed   "
        title="navigate to user section"
      >
        <span
          class="vtmx-user-line "
          data-testid="react--components--sidebar--menu--button--icon"
          style="font-size: 24px;"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--collapsed   "
      title="navigate to user section"
    >
      <span
        class="vtmx-user-line "
        data-testid="react--components--sidebar--menu--button--icon"
        style="font-size: 24px;"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`uiSidebarMenuButton > snapshot testing > when the menu is not collapsed > should match the not collapsed button snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--uncollapsed   "
        title="navigate to user section"
      >
        <span
          class="vtmx-user-line "
          data-testid="react--components--sidebar--menu--button--icon"
          style="font-size: 24px;"
        />
        <span
          class="typo_text-3 ml-4"
        >
          user button
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--uncollapsed   "
      title="navigate to user section"
    >
      <span
        class="vtmx-user-line "
        data-testid="react--components--sidebar--menu--button--icon"
        style="font-size: 24px;"
      />
      <span
        class="typo_text-3 ml-4"
      >
        user button
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
